File: ui/visual-component/src/main/java/org/visual/component/control/FileTreeView.java
package org.visual.component.control;

import java.io.File;
import java.io.IOException;
import java.nio.file.*;
import java.util.concurrent.CompletableFuture;
import javafx.application.Platform;
import javafx.beans.property.SimpleObjectProperty;
import javafx.scene.control.TreeItem;
import javafx.scene.control.TreeView;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class FileTreeView extends TreeView<File> {
  private final SimpleObjectProperty<File> _fileRoot = new SimpleObjectProperty<>();
  private final WatchService watcher = FileSystems.getDefault().newWatchService();

  public File getFileRoot() {
    return _fileRoot.get();
  }

  public void setFileRoot(File value) {
    if (value != null && !value.isDirectory()) {
      throw new IllegalArgumentException("Root must be a directory");
    }
    _fileRoot.set(value);
  }

  public FileTreeView() throws IOException {
    setShowRoot(false);
    _fileRoot.addListener(
        (observable, oldValue, newValue) -> {
          renderTree();
        });
  }

  private void renderTree() {
    TreeItem<File> rootItem = new TreeItem<>(_fileRoot.get());
    rootItem.setExpanded(true);
    createTreeItems(rootItem, _fileRoot.get());
    setRoot(rootItem);
  }

  private void createTreeItems(TreeItem<File> parentItem, File directory) {
    Path path = _fileRoot.get().toPath();
    try {
      path.register(
          watcher, StandardWatchEventKinds.ENTRY_CREATE, StandardWatchEventKinds.ENTRY_DELETE);
    } catch (IOException e) {
      e.printStackTrace();
    }
    File[] files = directory.listFiles();
    if (files != null) {
      for (File file : files) {
        TreeItem<File> treeItem = new TreeItem<>(file);
        parentItem.getChildren().add(treeItem);
        if (file.isDirectory()) {
          try {
            file.toPath()
                .register(
                    watcher,
                    StandardWatchEventKinds.ENTRY_CREATE,
                    StandardWatchEventKinds.ENTRY_DELETE);
          } catch (IOException e) {
            e.printStackTrace();
          }
          createTreeItems(treeItem, file);
        }
      }
    }
  }

  private void listen() {
    CompletableFuture.runAsync(
        () -> {
          while (true) {
            WatchKey key;
            try {
              key = watcher.take();
              for (WatchEvent<?> event : key.pollEvents()) {
                log.info("e:" + event.kind().type());
                log.info("e:" + event.context());
                Path context = (Path) event.context();
                if (context != null) {
                  Platform.runLater(this::renderTree);
                }
              }
              key.reset();
            } catch (InterruptedException e) {
              e.printStackTrace();
            }
          }
        });
  }
}

File: ui/visual-component/src/main/kotlin/org/visual/component/control/FileTreeView.kt
package org.visual.component.control

import io.github.oshai.kotlinlogging.KotlinLogging
import it.unimi.dsi.fastutil.objects.Object2IntSortedMaps
import java.nio.file.*
import javafx.application.Platform
import javafx.beans.property.SimpleObjectProperty
import javafx.beans.value.ObservableValue
import javafx.scene.control.TreeItem
import javafx.scene.control.TreeView
import kotlin.io.path.isDirectory
import lombok.extern.slf4j.Slf4j

@Slf4j
class FileTreeView : TreeView<Path>() {
  private val log = KotlinLogging.logger {}
  private val _fileRoot = SimpleObjectProperty<Path>()
  private val watcher: WatchService = FileSystems.getDefault().newWatchService()
  private val threadFactory = Thread.ofVirtual().name(this::class.simpleName, 0)
  private val activeCount by lazy { Object2IntSortedMaps.emptyMap<String>() }

  var fileRoot: Path?
    get() = _fileRoot.get()
    set(value) {
      require(!(value != null && !value.isDirectory())) { "Root must be a directory" }
      _fileRoot.set(value)
    }

  init {
    isShowRoot = false
    _fileRoot.addListener { _: ObservableValue<out Path?>?, _: Path?, _: Path? -> renderTree() }
  }

  private fun renderTree() {
    val rootItem = TreeItem(_fileRoot.get())
    rootItem.isExpanded = true
    createTreeItems(rootItem, _fileRoot.get())
    root = rootItem
  }

  private fun createTreeItems(parentItem: TreeItem<Path?>, directory: Path?) {
    val path = _fileRoot.get()
    path.register(
        watcher, StandardWatchEventKinds.ENTRY_CREATE, StandardWatchEventKinds.ENTRY_DELETE)
    val files = directory
    if (files != null) {
      //      for (file in files) {
      //        val treeItem = TreeItem(file)
      //        parentItem.children.add(treeItem)
      //        if (file.isDirectory) {
      //          try {
      //            file.toPath()
      //              .register(
      //                watcher,
      //                StandardWatchEventKinds.ENTRY_CREATE,
      //                StandardWatchEventKinds.ENTRY_DELETE
      //              )
      //          } catch (e: IOException) {
      //            e.printStackTrace()
      //          }
      //          createTreeItems(treeItem, file)
      //        }
      //      }
    }
  }

  private fun listen() {
    threadFactory.start {
      while (true) {
        var key: WatchKey
        try {
          key = watcher.take()
          for (event in key.pollEvents()) {
            log.atInfo { message = "e:" + event.kind().type() }
            log.atInfo { message = "e:" + event.context() }
            val context = event.context() as Path
            Platform.runLater { this.renderTree() }
          }
          key.reset()
        } catch (e: InterruptedException) {
          e.printStackTrace()
        }
      }
    }
  }
}
